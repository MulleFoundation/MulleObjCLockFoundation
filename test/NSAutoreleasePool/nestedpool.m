#ifndef __MULLE_OBJC__
# import <Foundation/Foundation.h>
# pragma message "this test does not work with Apple Foundation"
#else
# import <MulleObjC/MulleObjC.h>
#endif


@interface Foo : NSObject
@end

static int   n_instances;

@implementation Foo

+ (id) new
{
   ++n_instances;
   printf( "+new\n");
   return( [super new]);
}


- (void) dealloc
{
   --n_instances;
   printf( "-dealloc\n");
   [super dealloc];
}


@end



int  main()
{
   printf( "autoreleasepool 1 opens %d\n", n_instances);
   @autoreleasepool
   {
      [[Foo new] autorelease];

      printf( "autoreleasepool 2 opens %d\n", n_instances);
      @autoreleasepool
      {
         [[Foo new] autorelease];

         printf( "autoreleasepool 3 opens %d\n", n_instances);
         @autoreleasepool
         {
            [[Foo new] autorelease];
            printf( "autoreleasepool 3 closes %d\n", n_instances);
         }
         printf( "autoreleasepool 2 closes %d\n", n_instances);
      }
      printf( "autoreleasepool 1 closes %d\n", n_instances);
   }
}
