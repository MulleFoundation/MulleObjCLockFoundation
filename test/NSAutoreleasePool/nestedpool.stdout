autoreleasepool 1 opens 0
+new
autoreleasepool 2 opens 1
+new
autoreleasepool 3 opens 2
+new
autoreleasepool 3 closes 3
-dealloc
autoreleasepool 2 closes 2
-dealloc
autoreleasepool 1 closes 1
-dealloc
